### PR TITLE
Refactor Trim/Split/Split2KV functions in util.h

### PR DESCRIPTION
**Pull request by PragmaTwice**
- Solve #581
- Refactor Split and Split2KV
- Optimize some signature to const reference
- Add some unit test for string utils

### CODE REVIEW

**Review by tisonkun — COMMENTED** [review]
Generally looks good. One comment:

It seems that `Trim` will update the input string while `Split` doesn't. This looks inconsistent.

Also, if `Split` with empty delim, we may return the input string as is.

**Comment by PragmaTwice** [comment]
Thanks @tisonkun !

> `Trim` will update the input string while `Split` doesn't

`Trim` takes a copy of the first argument, not a reference.

> if `Split` with empty delim, we may return the input string as is.

done~

File: src/util.cc [hunk 1]
```diff
@@ -371,7 +371,14 @@ std::string Trim(std::string in, const std::string &chars) {
 std::vector<std::string> Split(const std::string &in, const std::string &delim) {
   std::vector<std::string> out;
 
-  if (in.empty() || delim.empty()) return out;
+  if (in.empty()) {
+    return out;
+  }
+
+  if (delim.empty()) {
```
Comment (tisonkun) [inline src/util.cc:378]: Sorry. After take a look at Java's `Spring.split` and C++20's [`split_view`](https://en.cppreference.com/w/cpp/ranges/split_view), its seems empty delim will cause output a string vector contains every character as a string member.

```java
jshell> "123".split("")
$1 ==> String[3] { "1", "2", "3" }

jshell> "1 2 3".split("")
$2 ==> String[5] { "1", " ", "2", " ", "3" }
```

```cpp
#include <iostream>
#include <iomanip>
#include <ranges>
#include <string_view>
 
int main() {
    constexpr std::string_view words{"Hello-_-C++-_-20-_-!"};
    constexpr std::string_view delim{""};
    for (const auto word : std::views::split(words, delim)) {
        std::cout << std::quoted(std::string_view(word.begin(), word.end())) << ' ';
    }
}
// "H" "e" "l" "l" "o" "-" "_" "-" "C" "+" "+" "-" "_" "-" "2" "0" "-" "_" "-" "!" 
// https://wandbox.org/permlink/5BZQHqUYd36t0WbQ
```

Comment (PragmaTwice) [reply src/util.cc:378]: So thanks, I'll fix it.
BTW, this function has some difference than some well known `split`: it split strings regarding to regex "a|b|c|..." where delim="abc...", so I am considering to add another split that split strings just regarding to the provided delimiter.

Comment (tisonkun) [reply src/util.cc:378]: Yes. We don't have to follow other system's design, but we should make our own utils semantic clear.

Comment (PragmaTwice) [reply src/util.cc:378]: done

File: src/util.cc [hunk 2]
```diff
@@ -359,42 +359,54 @@ std::string ToLower(std::string in) {
...
+  size_t begin = 0, end = in.find_first_of(delim);
   do {
-    pos = in.find_first_of(delim);
-    elem = in.substr(0, pos);
-    Trim(elem, delim, &trimed_elem);
-    if (!trimed_elem.empty()) out->push_back(trimed_elem);
-    in = in.substr(pos+1);
-  } while (pos != std::string::npos);
+    std::string elem = in.substr(begin, end - begin);
+    if (!elem.empty()) out.push_back(std::move(elem));
```
Comment (tisonkun) [inline src/util.cc:388]: `Split("1  2", " ")` should normally return `{"1", "", "2"}`. But with this condition, it returns `{"1", "2"}`.

I don't know if kvrocks depends on this behavior but point out it's diverged from other libraries.

Comment (PragmaTwice) [reply src/util.cc:388]: I think it seems more reasonable if `Split("1 2", " ")` returns `{"1", "2"}`, and it is same as previous behavior.
And `Split("1 2","")` returns `{"1", " ", "2"}`.

**Review by tisonkun — DISMISSED** [review]

File: src/util.cc (hunk 2 above)
Comment (git-hulk) [reply src/util.cc:388]: Yes, this `split` function has a bit different from most languages
since we want to ignore the empty value.

**Review by git-hulk — DISMISSED** [review]
Cool, this makes a lot sense.

File: src/util.cc [hunk 3]
```diff
@@ -538,25 +550,29 @@ std::string StringToHex(const std::string &input) {
...
 
+constexpr std::uint64_t ExpTo1024(size_t n) {
+  return static_cast<std::uint64_t>(1) << (10 * n);
+}
+
 void BytesToHuman(char *buf, size_t size, uint64_t n) {
   double d;
 
-  if (n < 1024) {
+  if (n < ExpTo1024(1)) {
```
Comment (ShooterIT) [inline src/util.cc]: actually this code is copied from redis.
but for this change, i don't feel it is clear, maybe we can define `KB MB GB TB PB`.

And this change is not related to theme.

Comment (PragmaTwice) [reply src/util.cc]: It just replace `1024*1024*...*1024` to something like `1024^n`, so I don't think it will affect readability. And it is not related to `bytes`.

> And this change is not related to theme

Agree. Maybe move it to a separate PR.

Comment (PragmaTwice) [reply src/util.cc]: These changes are reverted, I'll prepare a seperate PR to illustrate them.

**Review by tisonkun — APPROVED** [review]

**Review by ShooterIT — APPROVED** [review]

**Review by git-hulk — APPROVED** [review]

**Comment by PragmaTwice** [comment]
Thanks everyone! Merging...
